Add EIP: Bridged Token Standard (#7281)
ethereum/EIPs: EIPS/eip-7281.md

---
eip: 7281
title: Sovereign Bridged Token
description: An interface for creating fungible representations of tokens bridged across domains.
author: Shaito (@0xShaito), Excalibor (@excaliborr), Arjun Bhuptani (@arjunbhuptani)

> discussions-to: <URL> (TODO)

Reviewer: ```suggestion
discussions-to: https://ethereum-magicians.org/t/erc-7281-sovereign-bridged-tokens/14979
```

- Minimal Liquidity. Issuers want to minimize the costs and complexity of acquiring liquidity for their token on each supported bridge and chain. This property becomes increasingly important as the space eventually expands to 100s or 1000s of connected domains.

There has been some previous work to solve this problem. However, solutions have historically either failed to satisfy all of the above desirable properties or have been custom-built for only a single token ecosystem.

- Celer’s Open Canonical Token Standard proposed a lock-in free standard for bridging tokens to new domains. However, it largely targeted alternative L1s (that don’t already have a canonical bridge) and did not fully solve the fungibility problem. Regardless, Celer’s approach inspired some of the key thinking behind this standard.
- Maker’s Teleport facility allows for minting and burning canonical DAI between domains. While the approach solves for the desirable properties above, it is highly custom to Maker’s architecture and relies on Maker’s own economic security to function. Circle CCTP similarly solves this problem, but using a mechanism that only centralized token issuers can implement.

> - Token issuer multi-bridge implementations, e.g. Angle protocol, Frax Ferry, and Threshold Network’s tBTC. These examples solve for some or all of the above desiderata and can be applied more broadly to all tokens if coupled with minor additions for compatibility with existing deployed tokens.

Reviewer 2: > desiderata

Is this a typo, or a word I'm not smart enough to know?

Reviewer 3: de·sid·er·a·tum (noun) - plural noun: desiderata
**something that is needed or wanted.**
> "integrity was a desideratum"

- Maker’s Teleport facility allows for minting and burning canonical DAI between domains. While the approach solves for the desirable properties above, it is highly custom to Maker’s architecture and relies on Maker’s own economic security to function. Circle CCTP similarly solves this problem, but using a mechanism that only centralized token issuers can implement.
- Token issuer multi-bridge implementations, e.g. Angle protocol, Frax Ferry, and Threshold Network’s tBTC. These examples solve for some or all of the above desiderata and can be applied more broadly to all tokens if coupled with minor additions for compatibility with existing deployed tokens.

## Specification

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.

### Overview

The system proposed below has two main components:

- A standard interface for bridged tokens
- A wrapper (aka “Lockbox”) that allows existing tokens to adopt the above interface

### Token Interface

> All xERC-20 tokens MUST implement the standard ERC-20 interface.  Note that while many of the below functions are inspired by ERC777, implementers are NOT REQUIRED to adhere to the full ERC777 specification.

Reviewer 4: ```suggestion
All xERC-20 tokens MUST implement the standard ERC-20 interface.  Note that while many of the below functions are inspired by ERC-777, implementers are NOT REQUIRED to adhere to the full ERC-777 specification.
```

All EIP-XX tokens MUST implement the following interface.

```ts
interface IXERC-20 {
  /**
   * @notice Emits when a lockbox is set
   *
   * @param _lockbox The address of the lockbox
   */

  event LockboxSet(address _lockbox);

  /**
   * @notice Emits when a limit is set
   *
   * @param _mintingLimit The updated minting limit we are setting to the bridge
   * @param _burningLimit The updated burning limit we are setting to the bridge
   * @param _bridge The address of the bridge we are setting the limit too
   */

  event BridgeLimitsSet(uint256 _mintingLimit, uint256 _burningLimit, address indexed _bridge);

  /**
   * @notice Reverts when a user with too low of a limit tries to call mint/burn
   */

>   error IXERC-20_NotHighEnoughLimits();

Reviewer 4: ```suggestion
  error IXERC20_NotHighEnoughLimits();
```

Remove invalid syntax.

  /**
   * @notice Returns the current limit of a bridge
   *
   * @param _bridge The bridge we are viewing the limits of
   * @return _limit The limit the bridge has
   */

  function mintingCurrentLimitOf(address _bridge) external view returns (uint256 _limit);

  /**
   * @notice Returns the current limit of a bridge
   *
   * @param _bridge the bridge we are viewing the limits of
   * @return _limit The limit the bridge has
   */

  function burningCurrentLimitOf(address _bridge) external view returns (uint256 _limit);

  /**
   * @notice Mints tokens for a user
   * @dev Can only be called by a bridge
   * @param _user The address of the user who needs tokens minted
   * @param _amount The amount of tokens being minted
   */

  function mint(address _user, uint256 _amount) external;

  /**
   * @notice Burns tokens for a user
   * @dev Can only be called by a bridge
   * @param _user The address of the user who needs tokens burned
   * @param _amount The amount of tokens being burned
   */

  function burn(address _user, uint256 _amount) external;
}
```

Implementations MUST additionally satisfy the following requirements:

> - `mint` MUST check that the caller's current available `limit` is greater than or equal to `_amount`

Reviewer 4: ```suggestion

- `mint` MUST check that the caller's current available `limit` is greater than or equal to `_amount`
```

Insert blank line around list as required by validator

- `mint` MUST increase the supply of the underlying ERC-20 by `_amount` and reduce the current available `limit`
- `burn` MUST check that the caller's current available `limit` is greater than or equal to `_amount`
- `burn` MUST decrease the supply of the underlying ERC-20 by `_amount` and reduce the current available `limit`

### Lockbox

The lockbox tries to emulate the WETH contract interface as much as possible. Lockboxes MUST implement the following interface:

```ts

> interface IXERC-20Lockbox {

Reviewer 4: ```suggestion
interface IXERC20Lockbox {
```

Fix invalid interface definition syntax.

  /**
   * @notice Emitted when tokens are deposited into the lockbox
   */

  event Deposit(address _sender, uint256 _amount);

  /**
   * @notice Emitted when tokens are withdrawn from the lockbox
   */

  event Withdraw(address _sender, uint256 _amount);

  /**
   * @notice Reverts when a user tries to deposit native tokens on a non-native lockbox
   */

>   error IXERC-20Lockbox_NotNative();

Reviewer 4: ```suggestion
  error IXERC20Lockbox_NotNative();
```

Fix invalid interface definition syntax.

  /**
   * @notice Reverts when a user tries to deposit non-native tokens on a native lockbox
   */

  error IXERC-20Lockbox_Native();

  /**
   * @notice Reverts when a user tries to withdraw and the call fails
   */

>   error IXERC-20Lockbox_WithdrawFailed();

Reviewer 4: ```suggestion
  error IXERC20Lockbox_WithdrawFailed();
```

Fix invalid interface definition syntax.

  /**
   * @notice Deposit ERC-20 tokens into the lockbox
   *
   * @param _amount The amount of tokens to deposit
   */

  function deposit(uint256 _amount) external;

  /**
   * @notice Withdraw ERC-20 tokens from the lockbox
   *
   * @param _amount The amount of tokens to withdraw
   */

  function withdraw(uint256 _amount) external;
}
```

Lockboxes SHOULD additionally implement the following alternative `deposit` function for native (non-ERC-20) assets.

> ```ts

Reviewer 4: ```suggestion

```ts
```

Insert newline to make validator happy

    1. This problem exists in the current model if there is insufficient liquidity on the destination for a given bridge to complete a transaction within some defined slippage constraints. This case is not well-handled by bridges at the moment. Bridges are forced to either output some wrapped representation, or force users to wait until there is more liquidity. 
    2. In the limit-based approach, the bridge will similarly not be able to complete the transaction on the destination domain until it has more capacity available. However, a limit approach provides some more reasonable guarantees to the user: (1) user have a much higher degree of predictability around time and pricing of the outputted transaction, (2) users would not receive some wrapped representation, and (3) bridges would have a simpler pathway for users to send the tokens back to the source domain or any other destination.

### Aggregation
EIP-7281 introduces new dynamics for aggregators that massively improve user safety and experience across chains.

> There are two unsolved mechanism design problems that currently exist around bridge aggregation:

Reviewer 4: ```suggestion
There are two unsolved mechanism design problems that currently exist around bridge aggregation:

```

Add blank line around list

- **Arbitrum rollup bridge:**  Arbitrum allows token issuers to [write a custom gateway](https://developer.arbitrum.io/asset-bridging#other-flavors-of-gateways).
- **OP Stack bridge:** Optimism supports [adding a custom bridge](https://community.optimism.io/docs/guides/bridge-dev/#building-a-custom-bridge).
- **Polygon PoS bridge:** Polygon supports custom tokens [through their fx portal](https://wiki.polygon.technology/docs/develop/l1-l2-communication/fx-portal).
- **ZkSync bridge**: Zksync supports [custom bridges of any kind](https://era.zksync.io/docs/reference/concepts/bridging/bridging-asset.html#custom-bridges-on-l1-and-l2).
- **GnosisChain Omnibridge:** GnosisChain does not currently support custom tokens, but is in the process of redesigning their omnibridge and plans to allow this functionality.

Note: In most of the above cases, a custom bridge integration also means integration into canonical bridge UIs, ensuring that users have a consistent experience throughout the process. Additionally, a single xERC-20 custom bridge implementation could be built for each ecosystem and serve any number of token issuers.

## Reference Implementation

> You can find a reference implementation and associated test cases [here](https://github.com/defi-wonderland/xTokens/blob/dev/solidity/contracts/XERC-20.sol).

Reviewer 5: This link appears broken

```suggestion
You can find a reference implementation and associated test cases [here](https://github.com/defi-wonderland/xERC20/blob/dev/solidity/contracts/XERC20.sol).
```
